test/jdk/java/net/httpclient/AsyncShutdownNow.java

```diff
@@ -0,0 +1,452 @@
+            for (int i = 0; i < ITERATIONS; i++) {
+                URI uri = URI.create(uriString + "/concurrent/iteration-" + i);
+                HttpRequest request = HttpRequest.newBuilder(uri)
+                        .header("X-uuid", "uuid-" + requestCounter.incrementAndGet())
+                        .build();
+                out.printf("Iteration %d request: %s%n", i, request.uri());
+                CompletableFuture<HttpResponse<InputStream>> responseCF;
+                CompletableFuture<String> bodyCF;
+                final int si = i;
+                responseCF = client.sendAsync(request, BodyHandlers.ofInputStream())
+                        .thenApply((response) -> {
+                            out.println(si + ":  Got response: " + response);
+                            assertEquals(response.statusCode(), 200);
+                            return response;
+                        });
+                bodyCF = responseCF.thenApplyAsync(HttpResponse::body, readerService)
+                        .thenApply(AsyncShutdownNow::readBody)
+                        .thenApply((s) -> {
+                            assertEquals(s, MESSAGE);
+                            return s;
+                        });
+                long sleep = RANDOM.nextLong(5);
+                if (sleep > 0) {
+                    out.printf("%d: sleeping %d ms%n", i, sleep);
+                    Thread.sleep(sleep);
+                }
+                if (i == step) {
+                    out.printf("%d: shutting down client now%n", i, sleep);
+                    client.shutdownNow();
+                }
+                var cf = bodyCF.exceptionally((t) -> {
+                    Throwable cause = getCause(t);
+                    out.println(si + ": Got expected exception: " + cause);
+                    if (UncheckedIOException.class.isAssignableFrom(cause.getClass())) {
+                        if (cause.getCause() != null) {
+                            out.println(si + ": Got expected exception: " + cause);
+                            cause = cause.getCause();
+                        }
+                    }
+                    checkCause(String.valueOf(si), cause);
+                    return null;
+                });
+                bodies.add(cf);
+            }
+            CompletableFuture.allOf(bodies.toArray(new CompletableFuture<?>[0])).get();
+        } catch (Throwable throwable) {
+            failed = throwable;
+        } finally {
+            failed = cleanup(client, readerService, failed);
+        }
+        if (failed instanceof Exception ex) throw ex;
+        if (failed instanceof Error e) throw e;
+        assertTrue(client.isTerminated());
+    }
+
+    static Throwable cleanup(HttpClient client, ExecutorService readerService, Throwable failed) {
+        try {
+            try {
+                if (client.awaitTermination(Duration.ofMillis(2000))) {
```

Comment:
Hello Daniel, given our past experience of such timeouts leading to intermittent failures (on slow systems), do you think we should perhaps use a timeout duration of one day here? That way we don't have to guess the right timeout value and the jtreg test framework can timeout this test if it executes past the configured test timeout.

Reply:
If it fails I'd like to do some analysis of the logs before just increasing the timeout.

Reply:
That sounds reasonable.